### PR TITLE
[LeoFS] Added Multipart Upload Function

### PR DESCRIPTION
**Pull request by windkit**

### CODE REVIEW

File: src/basho_bench_driver_leofs.erl [hunk 1]
```diff
@@ -434,6 +463,10 @@ do_put(Url, Headers, Value, RetryOnOL) ->
             {error, Reason}
     end.
 
+do_mp(Url, Headers, Size, PartSize, PartBin, RetryOnOL) ->
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: this function is doing nothing at all

File: src/basho_bench_driver_leofs.erl [hunk 2]
```diff
@@ -678,3 +711,132 @@ normalize_error(Method, {'EXIT', {timeout, _}})  -> {error, {Method, timeout}};
...
+        {ok, Code, _Header, _Body} ->
+            {error, {http_error, Code}};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+upload_parts(Url, Headers, UploadId, Size, PartSize, PartBin, RetryOnOL) ->
+    upload_parts(Url, Headers, UploadId, Size, PartSize, PartBin, RetryOnOL, 1, <<>>).
+
+upload_parts(Url, Headers, UploadId, Size, PartSize, PartBin, RetryOnOL, Acc, _)
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: The last argument seems to be not used.

File: src/basho_bench_driver_leofs.erl [hunk 3]
```diff
@@ -678,3 +711,132 @@ normalize_error(Method, {'EXIT', {timeout, _}})  -> {error, {Method, timeout}};
...
+                                          integer_to_list(Acc),
+                                          "&uploadId=",
+                                          UploadId
+                                         ])},
+    TS = leo_date:now(),
+    Headers_2 = [
+                 {"Date", leo_http:rfc1123_date(TS)},
+                 {"content-type", ?S3_CONTENT_TYPE},
+                 {"authorization", gen_sig("PUT", Url_2, TS)}
+                ],
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: The part of generating Url_2 and Headers_2 can be commonized with the later.

File: src/basho_bench_driver_leofs.erl [hunk 4]
```diff
@@ -678,3 +711,132 @@ normalize_error(Method, {'EXIT', {timeout, _}})  -> {error, {Method, timeout}};
...
+    case send_request(Url_2, Headers ++ Headers_2,
+                       put, Bin, [{response_format, binary}]) of
+        {ok, "200", _Header, _Body} ->
+            complete_mp(Url, Headers, UploadId, RetryOnOL, Acc + 1, PartBinETag, erlang:md5(Bin));
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: Incrementing Count cause the unnecessary decrement when calling gen_mp_entries for the first time.

File: src/basho_bench_driver_leofs.erl [hunk 5]
```diff
@@ -678,3 +711,132 @@ normalize_error(Method, {'EXIT', {timeout, _}})  -> {error, {Method, timeout}};
...
+        {ok, "204", _Header, _Body} ->
+            complete_mp(Url, Headers, UploadId, RetryOnOL, Acc + 1, PartBinETag, erlang:md5(Bin));
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: Same with the above.

File: src/basho_bench_driver_leofs.erl [hunk 6]
```diff
@@ -678,3 +711,132 @@ normalize_error(Method, {'EXIT', {timeout, _}})  -> {error, {Method, timeout}};
...
+        {ok, "503", _Header, _Body} ->
+            retry_handler(fun() -> upload_parts(Url, Headers, UploadId, Size, PartSize, PartBin, RetryOnOL, Acc, <<>>) end, RetryOnOL);
+        {ok, Code, _Header, _Body} ->
+            {error, {http_error, Code}};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+complete_mp(Url, Headers, UploadId, RetryOnOL, Count, PartBinETag, LastETag) ->
+    EntriesBin = gen_mp_entries(1, Count - 1, PartBinETag, LastETag, <<>>),
```
Comment (mocchira) [inline src/basho_bench_driver_leofs.erl]: This decrement get to be unnecessary when the above increments are fixed.

**Comment by windkit** [comment]
@mocchira I have updated the PR, thanks.

**Pull request merged**